tests/ArrayHelperTest.php

```diff
@@ -1263,4 +1263,37 @@ public function testGetObjectVars()
             'content' => 'test'
         ], ArrayHelper::getObjectVars(new Post2()));
     }
+
+    public function testNestedPathStaysNestedAfterSet()
+    {
+        $array = ['a' => ['b' => ['c' => 'old']]];
+
+        $value = ArrayHelper::getValue($array, 'a.b.c');
+        $this->assertEquals('old', $value);
+
+        ArrayHelper::setValue($array, 'a.b.c', 'new');
+
+        $value = ArrayHelper::getValue($array, 'a.b.c');
+        $this->assertEquals('new', $value);
+
+        $this->assertEquals(['a' => ['b' => ['c' => 'new']]], $array);
+    }
+
+    /**
+     * https://github.com/yiisoft/arrays/issues/48
+     */
+    public function testFlatPathStaysFlatAfterSet()
+    {
+        $array = ['a.b.c' => 'old'];
+
+        $value = ArrayHelper::getValue($array, 'a.b.c');
+        $this->assertEquals('old', $value);
+
+        ArrayHelper::setValue($array, 'a.b.c', 'new');
+
+        $value = ArrayHelper::getValue($array, 'a.b.c');
+        $this->assertEquals('new', $value);
+
+        $this->assertEquals(['a.b.c' => 'new'], $array);
+    }
```

Comment:
```suggestion
    /**
     * https://github.com/yiisoft/arrays/issues/48
     *
     * @dataProvider compositeKeysProvider
     */
    public function testFlatPathStaysFlatAfterSet(array $initial, array $resultant)
    {
        $value = ArrayHelper::getValue($initial, 'a.b.c');
        $this->assertEquals('old', $value);

        ArrayHelper::setValue($initial, 'a.b.c', 'new');

        $value = ArrayHelper::getValue($initial, 'a.b.c');
        $this->assertEquals('new', $value);

        $this->assertEquals($resultant, $initial);
    }

    public function compositeKeysProvider(): array
    {
        return [
            [
                ['a' => ['b' => ['c' => 'old']]],
                ['a' => ['b' => ['c' => 'new']]],
            ],
            [
                ['a.b' => ['c' => 'old']],
                ['a.b' => ['c' => 'new']],
            ],
            [
                ['a' => ['b.c' => 'old']],
                ['a' => ['b.c' => 'new']],
            ],
            [
                ['a.b.c' => 'old'],
                ['a.b.c' => 'new'],
            ],
        ];
    }
```

Reply:
2nd and 3rd cases are not valid.

For keys with both nesting and dots, you must use array path syntax, not dot syntax.